### PR TITLE
Remove dtype from `chainerx.Scalar`

### PR DESCRIPTION
**Pull request by asi1024**

### CODE REVIEW

**Comment by okapies** [comment]
It seems this declaration should also be removed from the type stub. https://github.com/chainer/chainer/blob/8ea6741ecfd19b7ae58071c94a4410fc0c87da65/chainerx/__init__.pyi#L237-L238

File: chainerx_cc/chainerx/dtype.h [hunk 1]
```diff
@@ -224,6 +224,22 @@ inline const char* GetDtypeName(Dtype dtype) {
 // Gets the dtype of given name.
 Dtype GetDtype(const std::string& name);
 
+// Returns the default dtype.
+inline Dtype GetDefaultDtype(DtypeKind kind) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/dtype.h]: Does this have to be public?
How about `internal` in `routines/type_utils.h`?

File: chainerx_cc/chainerx/routines/creation_test.cc [hunk 2]
```diff
@@ -623,14 +623,6 @@ TEST_P(CreationTest, ArangeEmpty) {
     });
 }
 
-TEST_P(CreationTest, ArangeNoDtype) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/creation_test.cc]: Why is this test being removed?

File: chainerx_cc/chainerx/routines/creation_test.cc [hunk 3]
```diff
@@ -101,7 +101,7 @@ class CreationTest : public ::testing::TestWithParam<std::string> {
             Array x = Full(Shape{3, 2}, scalar);
             EXPECT_NE(x.data(), nullptr);
             EXPECT_EQ(x.shape(), Shape({3, 2}));
-            EXPECT_EQ(x.dtype(), scalar.dtype());
+            EXPECT_EQ(GetKind(x.dtype()), scalar.kind());
```
Comment (niboshi) [inline chainerx_cc/chainerx/routines/creation_test.cc]: This check was testing the value of `x.dtype()`. We shouldn't apply `GetKind`.

File: chainerx_cc/chainerx/scalar.h [hunk 4]
```diff
@@ -62,20 +48,15 @@ class Scalar {
...
-    Dtype dtype() const { return dtype_; }
+    DtypeKind kind() const { return kind_; }
 
     std::string ToString() const;
 
     bool operator==(Scalar other) const {
-        // TODO(niboshi): Support different dtypes
-        if (dtype_ != other.dtype_) {
-            return false;
+        if (this->kind() == DtypeKind::kFloat || other.kind() == DtypeKind::kFloat) {
```
Comment (niboshi) [inline chainerx_cc/chainerx/scalar.h]: Add test for this behavioral change.

File: chainerx_cc/chainerx/testing/array_check.h [hunk 5]
```diff
@@ -101,16 +101,19 @@ void ExpectDataEqual(Container&& expected_data, const Array& actual) {
...
-        if (IsNan(expected)) {
-            EXPECT_TRUE(IsNan(actual_value)) << "where i is " << it.raw_index();
-        } else {
-            EXPECT_EQ(expected, actual_value) << "where i is " << it.raw_index();
+    VisitDtype(actual.dtype(), [&](auto pt) {
+        using U = typename decltype(pt)::type;
+        IndexableArray<const U> actual_iarray{native_actual};
+        Indexer<> indexer{actual.shape()};
+        for (auto it = indexer.It(0); it; ++it) {
+            T actual_value = actual_iarray[it];
```
Comment (niboshi) [inline chainerx_cc/chainerx/testing/array_check.h]: >`T actual_value`

Is this correct?

File: tests/chainerx_tests/unit_tests/routines_tests/test_creation.py [hunk 6]
```diff
@@ -45,6 +44,19 @@ def _is_all_finite(obj):
         return numpy.isfinite(obj)
 
 
+def cast_to_expected_dtype_if_numpy_array(xp, array, dtype=None):
+    assert isinstance(array, xp.ndarray)
+    assert xp in (numpy, chainerx)
+    if xp is chainerx or dtype is not None:
+        return array
+    assert array.dtype.char in ('?', 'l', 'd')
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_creation.py]: Do not use dtype char codes. NumPy discourages it.

File: tests/chainerx_tests/unit_tests/routines_tests/test_creation.py [hunk 7]
```diff
@@ -594,18 +613,6 @@ def test_full_with_dtype(xp, shape, dtype_spec, value, device):
     return xp.full(shape, value, dtype_spec)
 
 
-@pytest.mark.parametrize(
-    'value', [True, False, -2, 0, 1, 2, 2.3, float('inf'), float('nan')])
-@pytest.mark.parametrize_device(['native:0', 'cuda:0'])
-def test_full_with_scalar(shape, dtype, value, device):
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_creation.py]: Why is this test being removed?

File: tests/chainerx_tests/unit_tests/routines_tests/test_creation.py [hunk 8]
```diff
@@ -45,6 +44,19 @@ def _is_all_finite(obj):
         return numpy.isfinite(obj)
 
 
+def cast_to_expected_dtype_if_numpy_array(xp, array, dtype=None):
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_creation.py]: Why not use `dtype_utils.cast_if_numpy_array`?

File: tests/chainerx_tests/unit_tests/routines_tests/test_creation.py [hunk 9]
```diff
@@ -879,7 +892,7 @@ def test_eye_invalid_NMk_type(xp, N, M, k, device):
 @pytest.mark.parametrize('transpose', [False, True])
 @pytest.mark.parametrize_device(['native:0', 'cuda:0'])
 def test_diag(xp, k, shape, transpose, device):
-    v = xp.arange(array_utils.total_size(shape)).reshape(shape)
+    v = xp.arange(array_utils.total_size(shape), dtype='i').reshape(shape)
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_creation.py]: Do not use char codes.

File: tests/chainerx_tests/unit_tests/routines_tests/test_creation.py [hunk 10]
```diff
@@ -891,7 +904,7 @@ def test_diag(xp, k, shape, transpose, device):
 @pytest.mark.parametrize('shape', [(), (2, 1, 2), (2, 0, 1)])
 @pytest.mark.parametrize('device', ['native:1', 'native:0'])
 def test_diag_invalid_ndim(xp, k, shape, device):
-    v = xp.arange(array_utils.total_size(shape)).reshape(shape)
+    v = xp.arange(array_utils.total_size(shape), dtype='i').reshape(shape)
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_creation.py]: Ditto.

File: tests/chainerx_tests/unit_tests/routines_tests/test_creation.py [hunk 11]
```diff
@@ -901,7 +914,7 @@ def test_diag_invalid_ndim(xp, k, shape, device):
 @pytest.mark.parametrize('shape', [(), (4,), (2, 3), (6, 5), (2, 0)])
 @pytest.mark.parametrize_device(['native:0', 'cuda:0'])
 def test_diagflat(xp, k, shape, device):
-    v = xp.arange(array_utils.total_size(shape)).reshape(shape)
+    v = xp.arange(array_utils.total_size(shape), dtype='i').reshape(shape)
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_creation.py]: Ditto.

File: tests/chainerx_tests/unit_tests/routines_tests/test_creation.py [hunk 12]
```diff
@@ -911,7 +924,7 @@ def test_diagflat(xp, k, shape, device):
 @pytest.mark.parametrize('shape', [(2, 1, 2), (2, 0, 1)])
 @pytest.mark.parametrize('device', ['native:1', 'native:0'])
 def test_diagflat_invalid_ndim(xp, k, shape, device):
-    v = xp.arange(array_utils.total_size(shape)).reshape(shape)
+    v = xp.arange(array_utils.total_size(shape), dtype='i').reshape(shape)
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_creation.py]: Ditto.

File: tests/chainerx_tests/unit_tests/routines_tests/test_math.py [hunk 13]
```diff
@@ -62,17 +62,10 @@ def test_add_scalar(scalar, device, shape, dtype):
     expected = x_np + numpy.dtype(dtype).type(scalar)
 
     x = chainerx.array(x_np)
-    scalar_chx = chainerx.Scalar(scalar, dtype)
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_math.py]: Why are these tests being removed?

File: tests/chainerx_tests/unit_tests/test_scalar.py [hunk 14]
```diff
@@ -15,114 +15,19 @@ def _check_cast_scalar_equals_data(scalar, data):
...
-    (float('inf'), chainerx.float64),
-    (float('nan'), chainerx.float64),
-    (True, chainerx.bool_),
-    (False, chainerx.bool_),
+@pytest.mark.parametrize('value', [
+    0, -1, 0x7fffffffffffffff, -0x8000000000000000, 0.0,
+    float('inf'), float('nan'), True, False,
 ])
-def test_init_without_dtype(value, dtype):
+def test_init_without_dtype(value):
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/test_scalar.py]: How about renaming to `test_init`?

Comment (niboshi) [inline tests/chainerx_tests/unit_tests/test_scalar.py]: How about renaming to `test_init`?

File: tests/chainerx_tests/unit_tests/test_scalar.py [hunk 15]
```diff
@@ -200,15 +105,7 @@ def test_cast(value):
...
-
-    if isinstance(value, bool):
-        assert scalar.dtype == chainerx.bool_
-    elif isinstance(value, int):
-        assert scalar.dtype == chainerx.int64
-    elif isinstance(value, float):
-        assert scalar.dtype == chainerx.float64
-    else:
-        assert False
+    assert isinstance(scalar.tolist(), type(value))
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/test_scalar.py]: Should this check be done in this test?

Comment (niboshi) [inline tests/chainerx_tests/unit_tests/test_scalar.py]: How aboud adding a test to check `kind`?

Comment (niboshi) [reply tests/chainerx_tests/unit_tests/test_scalar.py]: (discussed) Rename this test to `test_tolist`.

File: tests/chainerx_tests/unit_tests/routines_tests/test_creation.py [hunk 16]
```diff
@@ -45,6 +46,16 @@ def _is_all_finite(obj):
...
 
 
+def _get_default_dtype(value):
+    if isinstance(value, bool):
+        return 'bool_'
+    if isinstance(value, int):
+        return 'int32'
+    if isinstance(value, float):
+        return 'float32'
+    raise ValueError('There is no default dtype for {}'.format(type(value)))
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_creation.py]: `assert` is better (this is a test logic failure)

File: tests/chainerx_tests/unit_tests/routines_tests/test_creation.py [hunk 17]
```diff
@@ -595,12 +615,12 @@ def test_full_with_dtype(xp, shape, dtype_spec, value, device):
...
 @pytest.mark.parametrize(
-    'value', [True, False, -2, 0, 1, 2, 2.3, float('inf'), float('nan')])
+    'value', [True, False, -2, 0, 1, 2, 2.5, float('inf'), float('nan')])
 @pytest.mark.parametrize_device(['native:0', 'cuda:0'])
 def test_full_with_scalar(shape, dtype, value, device):
-    scalar = chainerx.Scalar(value, dtype)
+    scalar = chainerx.Scalar(value)
     a = chainerx.full(shape, scalar)
-    if scalar.dtype.kind == 'f' and math.isnan(float(scalar)):
+    if isinstance(scalar.tolist(), float) and math.isnan(float(scalar)):
```
Comment (niboshi) [inline tests/chainerx_tests/unit_tests/routines_tests/test_creation.py]: `scalar.tolist()` -> `value`?
`float(scalar)` -> `value`?

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
LGTM

**Pull request merged**